### ISSUE TITLE
some codes to check distance calculations

### ISSUE THREAD

**Issue by qodesign** [esabo/CodingTheory#2]

**Comment by esabo** [comment]
Why was this closed?

**Comment by qodesign** [comment]
I wanted to spare you the trouble of reading in the codes in my format. Looking at the list of quantum codes in your library you will probably get more coverage from these than the small example I uploaded.

**Comment by esabo** [comment]
`using DelimitedFiles
QR = readdlm("test/QR.txt",  Int);
QRstabs = QR[26:end-1, :];
using Nemo
F, _ = FiniteField(2, 1, "α");
QRstabsF = matrix(F, QRstabs);
QRQC = QuantumCode(QRstabsF, true)
[[25, 1]]_2 CSS code.

X-stabilizer matrix: 12 × 25
         chi(0) 1 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 0 0 0 0 0 0 0
         chi(0) 0 1 0 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0 0 0 0
         chi(0) 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0
         chi(0) 0 0 0 1 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 1 1 1 1
         chi(0) 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 0 0 0 0
         chi(0) 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 1 0 1 1 0 0 0 0
         chi(0) 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0
         chi(0) 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 1 0 0 0 1 1 1 1
         chi(0) 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 1 1 0 1 1 1
         chi(0) 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 1 0 0 1 1
         chi(0) 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 1
         chi(0) 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1 1 1 1 1
Z-stabilizer matrix: 12 × 25
         chi(0) 1 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0
         chi(0) 1 1 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0
         chi(0) 1 1 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0
         chi(0) 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0
         chi(0) 1 1 0 0 1 1 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0
         chi(0) 1 1 1 0 1 1 1 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0
         chi(0) 0 0 0 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0
         chi(0) 0 0 0 0 1 1 0 0 1 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0
         chi(0) 0 0 0 0 1 1 1 0 1 1 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0
         chi(0) 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 1
         chi(0) 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0 1
         chi(0) 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 1 1
logicals(QRQC)
1-element Vector{Tuple{fq_nmod_mat, fq_nmod_mat}}:
([0 0 0 1 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0], [0 0 0 ω 0 0 0 ω 0 0 0 ω 0 0 0 0 0 0 0 0 0 ω 0 0 0])
`
Easy enough. I'm just still fixing some small bugs before verifying what I got for the distances. Basically I have two functions: 1) build graph and do and 2) build equivalent but smaller graph and do. My first function always works. The second function sometimes works and has a slight bug. The key is that your codes have a slightly complicated internal structure to them which make the graph and computation time large. So if I can't reduce it properly for the n = 25 codes, I shouldn't try it on the larger codes yet.

**Comment by qodesign** [comment]
I didn't realize you actually started looking at this. All four codes are in "standard" or "canonical" stabilizer form.

Actually you can define multiple codes from a matrix like this : for example you can define a [[n=25,k=2,d=?]] code by taking 13 rows as stabilizers/destabilizers and 2 rows as logicalz/logicalx (instead of 14 & 1), but that's an experiment for later.

I started reading your paper; I'm familiar with the trellises and Viterbi algorithm in the context of classical coding; so some parts look familiar. I'm curious to see how you handle paths that are in the stabilizer vs paths in the stabilizer + centralizer. Once your code is ready I'll see if I can experiment with it...I have to warm up to the idea of using Julia (I've never used it before).

(I can reopen this issue if you like)

**Comment by esabo** [comment]
The second and third follow up papers contain the useful parts for this. There are multiple ways to handle the set difference. The way I’m doing it at this very second is to subtract the weight enumerator of the stabilizer from that of the normalizer. In terms of that paper though, these trellises are sometimes large because the span length of the generators are high. If a permutation was applied to reduce that, it would go considerably faster.

**Comment by qodesign** [comment]
the weight enumerator can explode in size for larger codes. I guess Julia has built in functions to handle huge integers. C needs gmp (gnu multi precision) library to calculate weight enumerators even for modest size codes.

**Comment by esabo** [comment]
Definitely, it's merely what I was working on when you posted. The long term solution won't do that. I imagine there being several if statements for user options. For quantum it's not as bad though because 1) I am using the complete weight enumerator instead of Hamming, so the coefficients are a bit more spread out, 2) the objects these are generally considered computable in finite time for are small enough not to run into too many problems, 3) you can tweak the stabilizer signs to cut the coefficients into pieces as well.

**Comment by qodesign** [comment]
it looks like there's an update with the 4 [[25,1,d]] codes included in the test directory.
Where you able to calculate distance for all 4? What do you get?
